src/main/kotlin/io/perfometer/runner/CoroutinesScenarioRunner.kt

```diff
@@ -0,0 +1,62 @@
+package io.perfometer.runner
+
+import io.perfometer.dsl.HttpStep
+import io.perfometer.dsl.PauseStep
+import io.perfometer.dsl.RequestStep
+import io.perfometer.http.client.HttpClient
+import io.perfometer.statistics.ConcurrentQueueScenarioStatistics
+import io.perfometer.statistics.PauseStatistics
+import io.perfometer.statistics.RequestStatistics
+import io.perfometer.statistics.ScenarioStatistics
+import kotlinx.coroutines.*
+import java.time.Duration
+import java.time.Instant
+
+internal class CoroutinesScenarioRunner(
+        private val httpClient: HttpClient,
+        private val scenarioStatistics: ScenarioStatistics = ConcurrentQueueScenarioStatistics(Instant.now())
+) : ScenarioRunner {
+
+    override fun runUsers(userCount: Int, block: () -> Unit) {
+        runBlocking {
+            val jobs: List<Job> = (1..userCount).map {
+                launch {
+                    block()
+                }
+            }
+            jobs.joinAll()
+        }
+    }
+
+    override fun runStep(step: HttpStep) {
+        when (step) {
+            is RequestStep -> executeHttp(step)
+            is PauseStep -> pauseFor(step.duration)
+        }
+    }
+
+    private fun pauseFor(duration: Duration) {
+        runBlocking {
+            delay(duration.toMillis())
+        }
+        scenarioStatistics.gather(PauseStatistics(duration))
+    }
+
+    private fun executeHttp(requestStep: RequestStep) {
```

Comment:
`executeHttp` method is exactly the same as in the `ThreadPoolScenarioRunner`, maybe it would be a good idea to have it separated a dedicated class using composition here? 

Also, I see that my initial implementation was bad, as the runner class has too many responsibilities: if for some reason the logic of executing requests and gathering statistics change, the runner(s) will also have to change. This class should be responsible for providing, say, an 'infrastructure' - that is concurrency mechanism for running the steps, but not for their actual execution. The exception here may be pausing, as this is directly dependend on concurrency mechanism used. :thinking:

Reply:
At the moment I introduced base class to avoid repeating. It would be good to achieve better separation of concerns but executing HTTP should also be changed in the future for coroutines runner implementation to actually suspend coroutine, do not black a thread for IO wait.